accepted/0000-selective-versions-resolutions.md

```diff
@@ -0,0 +1,423 @@
+
+In this context, it is impossible to force the use of `typescript@2.3.2` for
+the whole project (except by flattening the whole project, which we don't want).
+
+It makes sense for typescript as the user intent is clearly to use typescript
+2.3.2 for compiling all its project, and with the current behaviour, the angular
+CLI (responsible of compiling `.ts` files) will simply use the 2.2.2 version
+from its `node_modules`.
+
+Similarly, even using such a content for `package.json`:
+```json
+ "devDependencies": {
+    "@angular/cli": "1.0.3"
+  }
+```
+
+The need could arise for forcing the use of `typescript@2.3.2` (or
+`typescript@2.1.0` for that matter).
+
+## Why?
+
+In these example, the need does not seem very important (the user could maybe 
+use `typescript@2.2.2` or ask the `@angular/cli` dev team to relax its
+constraints on typescript), but there could be cases where a nested dependency
+introduces a bug and the project developer would want to set a specific
+version for it (see for example this
+[comment](https://github.com/yarnpkg/yarn/issues/2763#issuecomment-302682844)).
+
+## Related scenario (out of scope of this document)
+
+An extension of this motivation is also the potential need for mapping nested
+dependencies to others. For example a project developer could want to map
+`typescript@>=2.0.0 <2.3.0` to `my-typescript-fork@2.0.0`.
+
+See alternatives solutions below also.
+
+# Detailed design
+
+The proposed solution is to make the `resolutions` field of the `package.json`
+file to be considered all the time and on a per-package basis (instead of
+only when the `--flat` parameter is used).
+
+When a nested dependency is being resolved by yarn, if the `resolutions` field
+contains a specification for this package, then it will be used instead.
+
+Special attention to the specific nature of package management in the npm
+ecosystem is given in this RFC: indeed, it is not unusual to have the same
+package being present as a nested dependency of multiple packages with
+different versions. It is thus possible within the `resolutions` field to
+express versions either for the whole dependency tree or only for a subset of
+it, using a syntax relying on glob patterns.
+
+Most of the examples are given with exact dependencies, but note that using a
+non-exact specification in the `resolutions` field should be accepted and
+resolved by yarn like it usually does. This subject is discussed below also.
+
+Any potentially counter-intuitive situation will result in a warning being
+issued. This subject is discussed at the end of this section.
+
+## Examples
+
+We the following packages and their dependencies:
+```
+package-a@1.0.0 and 2.0.0
+ |_ package-d1@1.0.0
+     |_ package-d2@1.0.0
+
+package-b@1.0.0
+ |_ package-d1@2.0.0
+
+package-c@1.0.0
+ |_ package-a@2.0.0
+     |_ ...
+```
+
+With:
+```json
+ "devDependencies": {
+    "package-a": "1.0.0",
+    "package-b": "1.0.0"
+  },
+  "resolutions": {
+    "**/package-d1": "3.0.0"

```

Comment:
please specify that version is always in the form of 'major.minor.patch' and not allows any wildcards like `3.1.x` or `3.1.*`.